# Calculate annual costs of electricity based on utility rates and outputed / modeled SAM load profile

## Utility rate plans mapping
https://www.pge.com/assets/pge/docs/account/rate-plans/residential-electric-rate-plan-pricing.pdf
https://www.pge.com/tariffs/en.html


In [1]:
# Baseline Allowance per Territory and Season
# https://www.pge.com/en/account/rate-plans/how-rates-work/baseline-allowance.html#accordion-2fb51186db-item-2ea52b55e4
# Baseline Allowances for E-TOU-C Rate Plan
baseline_allowances = {
    "PGE": {
        "E-TOU-C": {
            "territories": {
                "T": {  # Territory T
                    "summer": 6.5,  # kWh/day
                    "winter": 7.5,  # kWh/day
                },
                # Add other territories as needed
                "P": {
                    "summer": 13.5,
                    "winter": 11.0,
                },
                "Q": {
                    "summer": 9.8,
                    "winter": 11.0,
                },
                "R": {
                    "summer": 17.7,
                    "winter": 10.4,
                },
                "S": {
                    "summer": 15.0,
                    "winter": 10.2,
                },
                "V": {
                    "summer": 7.1,
                    "winter": 8.1,
                },
                "W": {
                    "summer": 19.2,
                    "winter": 9.8,
                },
                "X": {
                    "summer": 9.8,
                    "winter": 9.7,
                },
                "Y": {
                    "summer": 10.5,
                    "winter": 11.1,
                },
                "Z": {
                    "summer": 5.9,
                    "winter": 7.8,
                },
            }
        }
    }
}

In [4]:
rate_plans = {
    "PGE": {
        "E-TOU-C": {
            "summer": { 
                "peak": 0.49, 
                "offPeak": 0.39, 
                "peakHours": [16, 17, 18, 19, 20], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 21, 22, 23],
                "fixedCharge": 0.00,
                "baseline_allowance": baseline_allowances["PGE"]["E-TOU-C"]["territories"]["T"]["summer"],  # 6.5 kWh/day
            },
            "winter": { 
                "peak": 0.38, 
                "offPeak": 0.35, 
                "peakHours": [16, 17, 18, 19, 20], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 21, 22, 23],
                "fixedCharge": 0.00,
                "baseline_allowance": baseline_allowances["PGE"]["E-TOU-C"]["territories"]["T"]["winter"],  # 6.5 kWh/day
            }
        },
        "E-TOU-D": {
            "summer": { 
                "peak": 0.55, 
                "offPeak": 0.42, 
                "peakHours": [17, 18, 19], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 20, 21, 22, 23],
                "fixedCharge": 0.00,
            },
            "winter": { 
                "peak": 0.46, 
                "offPeak": 0.42, 
                "peakHours": [17, 18, 19], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 20, 21, 22, 23],
                "fixedCharge": 0.00,
            }
        },
        "EV2-A": {
            "summer": { 
                "peak": 0.62, 
                "offPeak": 0.31, 
                "partPeak": 0.51, 
                "peakHours": [16, 17, 18, 19, 20], 
                "partPeakHours": [15, 21, 22, 23], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                "fixedCharge": 0.00,
            },
            "winter": { 
                "peak": 0.49, 
                "offPeak": 0.31, 
                "partPeak": 0.48, 
                "peakHours": [16, 17, 18, 19, 20], 
                "partPeakHours": [15, 21, 22, 23], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                "fixedCharge": 0.00,
            }
        },
        "EV-B": {
            "summer": { 
                "peak": 0.69, 
                "offPeak": 0.33, 
                "partPeak": 0.44, 
                "peakHours": [14, 15, 16, 17, 18, 19, 20], 
                "partPeakHours": [7, 8, 9, 10, 11, 12, 13, 21, 22], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 23],
                "fixedCharge": 0.00,
            },
            "winter": { 
                "peak": 0.50, 
                "offPeak": 0.30, 
                "partPeak": 0.37, 
                "peakHours": [14, 15, 16, 17, 18, 19, 20], 
                "partPeakHours": [7, 8, 9, 10, 11, 12, 13, 21, 22], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 23],
                "fixedCharge": 0.00,
            }
        },
        "E-ELEC": {
            "summer": { 
                "peak": 0.60, 
                "offPeak": 0.38, 
                "partPeak": 0.44, 
                "peakHours": [16, 17, 18, 19, 20], 
                "partPeakHours": [15, 21, 22, 23], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                "fixedCharge": 15.00, # Only E-ELEC has fixed charges: https://www.pge.com/assets/pge/docs/account/rate-plans/residential-electric-rate-plan-pricing.pdf
            },
            "winter": { 
                "peak": 0.37, 
                "offPeak": 0.33, 
                "partPeak": 0.35, 
                "peakHours": [16, 17, 18, 19, 20], 
                "partPeakHours": [15, 21, 22, 23], 
                "offPeakHours": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14],
                "fixedCharge": 15.00,
            }
        }
    }
}

## Load SAM net load profile
How much needs to be imported from the grid

In [5]:
import pandas as pd

# Import load profile from SAM
df = pd.read_csv('alameda/load/alameda_typical_solarstorage_load_profiles_SAM.csv')

load_profile = df['Load Profile'].tolist()

df.head()

FileNotFoundError: [Errno 2] No such file or directory: 'alameda/load/alameda_typical_solarstorage_load_profiles_SAM.csv'

In [ ]:
from datetime import datetime, timedelta

def get_season(hour_index):
    """
    Determines the season (summer or winter) for a given hour index.
    
    Args:
        hour_index (int): Hour index (0 to 8759)
    
    Returns:
        str: 'summer' or 'winter'
    """
    start_date = datetime(year=2018, month=1, day=1)  # Year is same as NREL inputs
    current_datetime = start_date + timedelta(hours=hour_index)
    month = current_datetime.month
    
    if 6 <= month <= 9:
        return 'summer' # Summer = June 1 to Sept 30
    else:
        return 'winter' # Winter = Oct 1 to May 31
    

print(get_season(0)) # great it works

winter


In [ ]:
def calculate_annual_costs_enhanced(load_profile, rate_plans):
    """
    Enhanced version to calculate annual energy costs with per-season tracking.
    
    Args:
        load_profile (list or np.array): List of 8760 hourly load values in kWh.
        rate_plans (dict): Dictionary containing rate plan details.
    
    Returns:
        dict: Annual cost for each rate option.
    """
    from collections import defaultdict
    
    annual_costs = defaultdict(float) # Dictionary to store annual costs per rate plan
    monthly_max_demand = {plan: [0] * 12 for plan in rate_plans["PGE"]} # Dictionary to store monthly maximum demands per rate plan
    energy_costs_per_season = {plan: {"summer": 0.0, "winter": 0.0} for plan in rate_plans["PGE"]}  # Dictionary to store energy costs per season
    
    for hour_index in range(8760):
        season = get_season(hour_index)
        # Determine the month
        start_date = datetime(year=2023, month=1, day=1)
        current_datetime = start_date + timedelta(hours=hour_index)
        month = current_datetime.month  # 1 to 12
        month_idx = month - 1  # 0-based index for months
        
        for plan_name, plan_details in rate_plans["PGE"].items():
            season_rates = plan_details.get(season)
            if not season_rates:
                continue  # Skip if season not defined
            
            hour = current_datetime.hour
            if hour in season_rates["peakHours"]:
                rate = season_rates["peak"]
            elif "partPeakHours" in season_rates and hour in season_rates["partPeakHours"]:
                rate = season_rates["partPeak"]
            else:
                rate = season_rates["offPeak"]
            
            # Calculate energy cost for this hour
            energy_cost = load_profile[hour_index] * rate
            annual_costs[plan_name] += energy_cost
            energy_costs_per_season[plan_name][season] += energy_cost
            
            demand = load_profile[hour_index]  # kW
            if demand > monthly_max_demand[plan_name][month_idx]:
                monthly_max_demand[plan_name][month_idx] = demand
    
    # Apply fixed charges, TODO: real demand charges, taxes, and rebates
    total_annual_costs = {}
    
    for plan_name, plan_details in rate_plans["PGE"].items():
        total_fixed_charge = 0.0
        total_demand_charge = 0.0
        total_taxes = 0.0
        total_rebates = 0.0
        
        # Iterate through seasons
        for season in ["summer", "winter"]:
            season_rates = plan_details.get(season)
            if not season_rates:
                continue  # Skip if season not defined
            
            # Fixed Charge: $ per month
            fixed_charge = season_rates.get("fixedCharge", 0.0)
            total_fixed_charge += fixed_charge * 12  # Annual fixed charge
        
        # Total annual cost calculation
        energy_cost = annual_costs[plan_name]
        total_cost = energy_cost + total_fixed_charge + total_demand_charge + total_taxes - total_rebates
        total_annual_costs[plan_name] = total_cost
    
    return total_annual_costs

annual_costs = calculate_annual_costs_enhanced(load_profile, rate_plans)

# Print results
for plan, cost in annual_costs.items():
    print(f"Annual Cost for {plan}: ${cost:.2f}")

Annual Cost for E-TOU-C: $2298.75
Annual Cost for E-TOU-D: $2626.54
Annual Cost for EV2-A: $2470.53
Annual Cost for EV-B: $2623.46
Annual Cost for E-ELEC: $2691.89


# Sanity Check
# Check it against energy / electricity consumption
# CEC may have info about annual per capita electricity consumption

For the "total" load profile (without solar + storage):

Annual Cost for E-TOU-C: $2298.75
Annual Cost for E-TOU-D: $2626.54
Annual Cost for EV2-A: $2470.53
Annual Cost for EV-B: $2623.46
Annual Cost for E-ELEC: $2691.89


For the "grid imports" load profile:

Annual Cost for E-TOU-C: $856.95
Annual Cost for E-TOU-D: $1002.40
Annual Cost for EV2-A: $930.66
Annual Cost for EV-B: $892.98
Annual Cost for E-ELEC: $1205.96


https://www.eia.gov/outlooks/aeo/tables_ref.php

